### PR TITLE
✨ Basic signin with google

### PR DESCRIPTION
**Pull request by patheticGeek**

### CODE REVIEW

File: devsnest/Gemfile.lock [hunk 1]
```diff
@@ -172,6 +172,7 @@ GEM
       rake (>= 0.8.7)
       thor (>= 0.20.3, < 2.0)
     rake (13.0.1)
+    random-word (2.1.1)
```
Comment (vedansh) [inline devsnest/Gemfile.lock:175]: why is this required?

File: devsnest/Gemfile.lock [hunk 2]
```diff
@@ -276,6 +277,7 @@ DEPENDENCIES
   puma (~> 4.1)
   rack-cors (~> 1.1.1)
   rails (~> 6.0.3, >= 6.0.3.4)
+  random-word
```
Comment (vedansh) [inline devsnest/Gemfile.lock:280]: same?

Comment (patheticGeek) [reply devsnest/Gemfile.lock:280]: Forgot to remove that library, were trying to get random words in token instead of having an old fashioned random string

File: devsnest/app/controllers/api/v1/submissions_controller.rb [hunk 3]
```diff
@@ -10,15 +10,18 @@ def create
         if @bot.present?
           discord_id = params['data']['attributes']['discord_id']
           user = User.find_by(discord_id: discord_id)
+
+          return render_error('user_not_active_web') if user.nil? || !user.web_active
         else
           user = @current_user
         end
+
+        user_id = user.id
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/submissions_controller.rb:19]: extra

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 4]
```diff
@@ -59,18 +68,47 @@ def left_discord
...
       def login
         code = params['code']
-        user = User.fetch_discord_user(code)
-        if user.present?
-          sign_in(user)
-          set_current_user
-          return render_success(user.as_json.merge({ "type": 'users'})) if @current_user.present?
+        type = params['type']
+        
+        if type == 'discord'
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:78]: remove

Comment (vedansh) [reply devsnest/app/controllers/api/v1/users_controller.rb:78]: We need to sign in the user here, better just add a connect endpoint

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 5]
```diff
@@ -20,6 +20,12 @@ def group_id
         member = GroupMember.where(user_id: context[:user].id).first
         member.present? ? member.group_id : nil
       end
+
+      def bot_token
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/user_resource.rb:24]: can we add it to after_create

Comment (vedansh) [reply devsnest/app/resources/api/v1/user_resource.rb:24]: This will create a new token everytime

File: devsnest/db/schema.rb [hunk 6]
```diff
@@ -78,33 +68,22 @@
...
-    t.integer "mentor_id"
-    t.text "feedback"
-    t.integer "timeGiven"
-    t.integer "capability"
-    t.datetime "created_at", precision: 6, null: false
-    t.datetime "updated_at", precision: 6, null: false
-  end
-
-  create_table "mmts", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
+  create_table "scrums", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
```
Comment (vedansh) [inline devsnest/db/schema.rb:71]: This table should not exist here, can we recreate the schema file

File: devsnest/db/migrate/20210527180300_adding_bot_token_to_user.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddingBotTokenToUser < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :bot_token, :string, null: false, default: ''
+  end
```
Comment (vedansh) [inline devsnest/db/migrate/20210527180300_adding_bot_token_to_user.rb:4]: can we merge the above 2 migrations as they are updating the same table

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 8]
```diff
@@ -59,18 +68,47 @@ def left_discord
...
+            sign_in(user)
+            set_current_user
+            return render_success(user.as_json.merge({ "type": 'users'})) if @current_user.present?
+          end
+
+        end
+        return render_error({message: "Error occurred while authenticating"})
+      end
+      
+      def update_bot_token_to_google_user
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:104]: add user authentication

File: devsnest/app/models/user.rb [hunk 9]
```diff
@@ -5,27 +5,72 @@ class User < ApplicationRecord
          :jwt_authenticatable,
          jwt_revocation_strategy: JwtBlacklist
 
-  def self.fetch_discord_user(code)
-    token = fetch_access_token(code)
+  def self.fetch_discord_user(code,context)
```
Comment (vedansh) [inline devsnest/app/models/user.rb:8]: check if we need this function now

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 8 above)
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:104]: Let's also check, if user has web_active and discord_active as true, then we can ignore

File: devsnest/app/models/user.rb [hunk 10]
```diff
@@ -5,27 +5,72 @@ class User < ApplicationRecord
...
+      username: name,
+      email: email,
+      password: Devise.friendly_token[0, 20],
+      web_active: true,
+      image_url: avatar,
+      google_id: googleId
+    )
+  end
+
+  def self.create_discord_user(user_details,context)
```
Comment (vedansh) [inline devsnest/app/models/user.rb:61]: do we need this now?

File: devsnest/app/models/user.rb [hunk 11]
```diff
@@ -66,6 +110,11 @@ def self.fetch_user_details(token)
     response = HTTParty.post(url, :body => {}, :headers => headers)
     response.code == 200 ? JSON(response.read_body) : nil
   end
+  
+  def self.update_discord_id(user,temp_user)
+    user.discord_id = temp_user.discord_id
+    user.save
+    temp_user_id = temp_user.id
```
Comment (vedansh) [inline devsnest/app/models/user.rb:117]: temp_user.destroy

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 12]
```diff
@@ -59,18 +68,47 @@ def left_discord
...
-          set_current_user
-          return render_success(user.as_json.merge({ "type": 'users'})) if @current_user.present?
+        type = params['type']
+        
+        if type == 'discord'
+          user = User.fetch_discord_user(code,context)
+          if user.present? && @current_user == nil
+            sign_in(user)
+            set_current_user
+            byebug
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:83]: remove byebug

**Comment by vedansh** [comment]
@ayush893 we also have to write a small logic for submissions merge, while merging both the users

File: devsnest/app/controllers/api/v1/submissions_controller.rb (hunk 3 above)
Comment (vedansh) [reply devsnest/app/controllers/api/v1/submissions_controller.rb:19]: @ayush893 do we need this line

File: devsnest/app/models/user.rb [hunk 13]
```diff
@@ -66,6 +101,15 @@ def self.fetch_user_details(token)
...
     response.code == 200 ? JSON(response.read_body) : nil
   end
+  
+  def self.update_discord_id(user,temp_user)
+    user.discord_id = temp_user.discord_id
+    user.save
+    temp_user.destroy
+  end
+  def create_bot_token
+    user = self
```
Comment (vedansh) [inline devsnest/app/models/user.rb:111]: remove

File: devsnest/app/controllers/api/v1/submissions_controller.rb (hunk 3 above)
Comment (ayush893) [reply devsnest/app/controllers/api/v1/submissions_controller.rb:19]: we are doing submission by user_id whether he is bot_user or web_user. so if he submit from anywhere his submission will be done.

File: devsnest/app/models/user.rb (hunk 10 above)
Comment (ayush893) [reply devsnest/app/models/user.rb:62]: this one is now updating discord_id only.

File: devsnest/app/controllers/api/v1/submissions_controller.rb (hunk 3 above)
Comment (vedansh) [reply devsnest/app/controllers/api/v1/submissions_controller.rb:19]: but we are not using user_id, we are using user.id

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 14]
```diff
@@ -59,19 +68,44 @@ def left_discord
...
 
       def logout
-        render json: { notice: 'You logged out successfuly' }
+        render json: { notice: 'You logged out successfully' }
+      end
+
+      def connect_discord
+        code = params['code']
+        active_user = @current_user
+          user = User.fetch_discord_user(code,active_user)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:77]: looks like there are some issues with this, let's discuss this

**Comment by vedansh** [comment]
User.update_all(web_active:false)

File: devsnest/app/models/user.rb [hunk 15]
```diff
@@ -4,68 +4,102 @@ class User < ApplicationRecord
...
+  def self.update_discord_id(user, temp_user)
+    user.discord_id = temp_user.discord_id
+    Submission.merge_submission(temp_user, user)
+    user.save
+    temp_user.destroy
+  end
 
+  def create_bot_token
+    update(bot_token: Digest::SHA1.hexdigest([Time.now, rand].join))
+    save
```
Comment (vedansh) [inline devsnest/app/models/user.rb:103]: verify if this is required

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 16]
```diff
@@ -59,18 +67,47 @@ def left_discord
...
+
+          tmp_user = User.find_by(discord_id: discord_id)
+          return render_error({ message: 'Discord User is already connected to another user' }) if tmp_user.present? && tmp_user.web_active?
+
+          user = User.merge_discord_user(discord_id, @current_user)
+          if !user.discord_id.nil?
+            render_success(user.as_json.merge({ "type": 'users' }))
+          else
+            render_error({ message: 'Failed to connect discord' })
+          end
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:87]:
```
return render_error({ message: 'Failed to connect discord' }) if user.discord_id.nil?
return render_success(user.as_json.merge({ "type": 'users' }))
```

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 17]
```diff
@@ -5,8 +5,8 @@ module V1
     class UsersController < ApplicationController
       include JSONAPI::ActsAsResourceController
       before_action :simple_auth, only: %i[leaderboard report show]
-      before_action :bot_auth, only: %i[left_discord create index]
-      before_action :user_auth, only: [:logout, :me]
+      before_action :bot_auth, only: %i[left_discord create index get_token]
+      before_action :user_auth, only: %i[logout me]
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:9]: add connect_discord

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 18]
```diff
@@ -59,18 +67,47 @@ def left_discord
...
+          else
+            render_error({ message: 'Failed to connect discord' })
+          end
+        elsif params['data']['attributes']['bot_token'].present?
+          token = params['data']['attributes']['bot_token']
+          temp_user = User.find_by(bot_token: token)
+          return render_error({ message: 'Could not find user of provided token' }) if temp_user.nil?
+          return render_error({ message: 'Discord User is already connected to another user' }) if temp_user.web_active?
+
+          User.update_discord_id(@current_user, temp_user) if @current_user.present?
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:94]: remove if @current_user.present? and return render_success

File: package.json [hunk 19]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (vedansh) [inline package.json:1]: remove file

File: devsnest/app/models/user.rb (hunk 15 above)
Comment (sam0000-cmd) [reply devsnest/app/models/user.rb:103]: Not Required and removed

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 18 above)
Comment (sam0000-cmd) [reply devsnest/app/controllers/api/v1/users_controller.rb:94]: done

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 16 above)
Comment (sam0000-cmd) [reply devsnest/app/controllers/api/v1/users_controller.rb:87]: done

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 17 above)
Comment (sam0000-cmd) [reply devsnest/app/controllers/api/v1/users_controller.rb:9]: done

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 20]
```diff
@@ -59,18 +67,43 @@ def left_discord
...
+
+          user = User.merge_discord_user(discord_id, @current_user)
+          return render_error({ message: 'Failed to connect discord' }) if user.discord_id.nil?
+          return render_success(user.as_json.merge({ "type": 'users' }))
+        elsif params['data']['attributes']['bot_token'].present?
+          token = params['data']['attributes']['bot_token']
+          temp_user = User.find_by(bot_token: token)
+          return render_error({ message: 'Could not find user of provided token' }) if temp_user.nil?
+          return render_error({ message: 'Discord User is already connected to another user' }) if temp_user.web_active?
+          User.update_discord_id(@current_user, temp_user)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:90]: can you return with success if discord id is updated

Comment (sam0000-cmd) [reply devsnest/app/controllers/api/v1/users_controller.rb:90]: done

**Review by vedansh — APPROVED** [review]
